File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int,int>sumPrefix;
        int ans=0,sum=0,n=nums.size();
        sumPrefix[0]=0;

        for(int i=0;i<n;i++){
            sum+=nums[i];
            if(sum==k) ans++;
            if(sumPrefix.find(sum-k)!=sumPrefix.end()){
                ans+=sumPrefix[sum-k];
            }
            sumPrefix[sum]++;
        }
        return ans;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int target) {
        int n = nums.length;

        int[] prefix = new int[n];

        prefix[0] = nums[0];

        for (int i = 1; i < n; i++) {
            prefix[i] = prefix[i - 1] + nums[i];
        }

        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);

        int count = 0;
        for (int i = 0; i < n; i++) {
            if (map.containsKey(prefix[i] - target))
                count += map.get(prefix[i] - target);
            map.put(prefix[i], map.getOrDefault(prefix[i],0) + 1);
        }

        return count;
    }
}